# Homework 04

## 4.1 - Essay

I don't think that "deep" neural networks are going anywhere, and believe that
they're a big breakthrough that will be used for years in the future. They have much
more capability than things like expert systems and perceptrons, which ended
up being unsuccessful for bigger scale and more complicated problems.

There are a few reasons for this fact. First is the wide application of deep neural nets.
They have the capability of learning almost any type of problem, given that one
feeds the data in correctly and optimally. From image recognition (like the MNIST dataset), 
to text based input (like the IMDB dataset) there's a lot of areas where deep neural nets 
are very useful, and also widely successful. In our world today we also now fully have the capabilities to make
deep neural networks effective. We have tons and tons of data available, which is required so that our
networks become more generalized, and also learn things deeper and better. We also
have the computational power to train these networks on all this data, thanks to things
like Google's cloud computing and server farms. 

Overall, thanks the availability of Big data and lots of computational power, 
along with deep neural network's wide application and strong success I think
they're going to be used for years to come.

## 4.2 - Back-Propagation Computation

Initial Weights:

$\begin{aligned}
    &\begin{bmatrix}
    w_{i_1,h_1} & w_{i_1,h_2} \\
    w_{i_2,h_1} & w_{i_2,h_2}
    \end{bmatrix}
    \leftarrow
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} \\
    &\begin{bmatrix}
    w_{h_1, o_1} \\ 
    w_{h_2, o_1} 
    \end{bmatrix}
    \leftarrow
    \begin{bmatrix}
    0.14 \\
    0.15
    \end{bmatrix}
    \end{aligned}$
    

Input: (XOR: `[1, 1]` &rarr; `0`)
Compute the output

 $\begin{aligned}
    o_j &= 
    \begin{bmatrix}
    1 & 1 \\ 
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\
    0.15
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    1 * 0.11 + 1 * 0.21 & 1 * 0.12 + 1 * 0.08
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\ 
    0.15
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    0.32 & 0.20
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    0.32 * 0.14 + 0.20 * 0.15
    \end{bmatrix}
    \\ &= 0.0748 
    \end{aligned}
    \\
    $
    
Compute the error

 $\begin{aligned}
    L_2Error &= (0 - 0.0748)^2 \\
    &= 0.005595 \\
    \Delta_{o_1} &= (0 - 0.0748) \\
    &= -0.0748 \\
    \end{aligned}$
    
Back-propigation. Learning rate is 0.05, the activation function
is f(x) = x.

$\begin{aligned}
    \begin{bmatrix}
    w_{h_1, o_1} \\ 
    w_{h_2, o_1}
    \end{bmatrix} &\leftarrow 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} + 0.05 \cdot 
    \begin{bmatrix}
    0.32 \\ 
    0.20 
    \end{bmatrix} \cdot 1.0 \cdot  -0.0748 \\\\
    &= 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * 0.32 * 1.0 *  -0.0748 \\
    0.05 * 0.20 * 1.0 *  -0.0748 
    \end{bmatrix} \\
    &= 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} +
    \begin{bmatrix}
    -0.00120 \\
    -0.00748 
    \end{bmatrix} \\
    &=
    \begin{bmatrix}
    0.1388 \\ 
    0.1425
    \end{bmatrix}
    \end{aligned}$
    
    
  $\begin{aligned}
    \begin{bmatrix}
    w_{i_1,h_1} & w_{i_1,h_2} \\ 
    w_{i_2,h_1} & w_{i_2,h_2}
    \end{bmatrix} &\leftarrow 
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 0.05 \cdot
    \begin{bmatrix}
    1 & 1 \\ 
    1 & 1
    \end{bmatrix} \cdot 1.0 \odot
    \begin{bmatrix}
    0.14 & 0.15 \\ 
    0.14 & 0.15
    \end{bmatrix} \cdot -0.0748 \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * 1 * 1.0 & 0.05 * 1 * 1.0 \\ 
    0.05 * 1 * 1.0 & 0.05 * 1 * 1.0 \\ 
    \end{bmatrix} \odot 
    \begin{bmatrix}
    0.14 * -0.0748 & 0.15 * -0.0748\\ 
    0.14 * -0.0748 & 0.15 * -0.0748
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 & 0.05 \\ 
    0.05 & 0.05 
    \end{bmatrix} \odot 
    \begin{bmatrix}
    -0.0105 & -0.0112 \\
    -0.0105 & -0.0112
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * -0.0105 & 0.05 * -0.0112 \\
    0.05 * -0.0105 & 0.05 * -0.0112
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} +     
    \begin{bmatrix}
    -0.000525 & -0.00056 \\
    -0.000525 & -0.00056
    \end{bmatrix} \\ &= 
    \begin{bmatrix}
    0.109 & 0.119 \\
    0.2095 & 0.0794
    \end{bmatrix}  
    \end{aligned}$
    

## 4.3 - ConvNet

Loading in the fashion MNIST dataset

In [1]:
from keras.datasets import fashion_mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


Using TensorFlow backend.


Build the model.

In [2]:
from keras import layers
from keras import models

model = models.Sequential()

# Configure a convnet with 3 layers of convolutions and max pooling.
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Add layers to flatten the 2D image and then do a 10-way classification.
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)               

In [3]:
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=256)
model.evaluate(train_images, train_labels)

Epoch 1/5
60000/60000 [==============================] - 60s 1ms/step - loss: 0.7479 - accuracy: 0.7241

[0.27214640025993186, 0.9022666811943054]

In [4]:
model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 3s 300us/step

[0.3095951925754547, 0.8870000243186951]

The network architecture from the class example seemed to work the 
best for this problem. I tried adding layers, removing layers, 
changing the activation function, but nothing seemed to really make
a great improvement. 

There are a few things I changed in terms of compiling the model.
One thing was that I switched the optimizer from 
rmsprop to adam. The model using adam as an optimizer yielded better results. 
I increased my batch_size to 256, as that was yielding better results. 
Epoch size didn't actually seem to have a huge difference. Most of the models
I trained i trained on only 2 epochs. 

In the end I was able to achieve 88.7% accuracy on the test data.

